Refactor format_metadata_to_key

MVrachev opened the pull request:
**Fixes issue #**:

https://github.com/secure-systems-lab/securesystemslib/issues/220

**Description of the changes being introduced by the pull request**:

Refactor format_metadata_to_key so that it won't depend
on settings.HASH_ALGORITHMS.
There are couple of code snippets in tuf where in order to provide
a custom list of hash algorithms, tuf is temporary changing
settings.HASH_ALGORITHMS and then returning it to its formal state.

This could lead to problems in future and one should
be able to control hash algorithms used for generating keyids in f
ormat_metadata_to_key() without having to change
package level settings in securesystemslib.settings.

**Please verify and check that the pull request fulfils the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

MVrachev commented on securesystemslib/interface.py:
```diff
@@ -572,7 +572,7 @@ def import_ed25519_publickey_from_file(filepath):
   ed25519_key_metadata = securesystemslib.util.load_json_file(filepath)
   ed25519_key, junk = \
-    securesystemslib.keys.format_metadata_to_key(ed25519_key_metadata)
+    securesystemslib.keys.format_metadata_to_key(ed25519_key_metadata, securesystemslib.settings.HASH_ALGORITHMS)
```
Addressed in a follow on commit.

MVrachev commented on securesystemslib/interface.py:
```diff
@@ -817,7 +817,7 @@ def import_ecdsa_publickey_from_file(filepath):
   ecdsa_key_metadata = securesystemslib.util.load_json_file(filepath)
   ecdsa_key, junk = \
-    securesystemslib.keys.format_metadata_to_key(ecdsa_key_metadata)
+    securesystemslib.keys.format_metadata_to_key(ecdsa_key_metadata, securesystemslib.settings.HASH_ALGORITHMS)
```
Addressed in a follow on commit.

MVrachev commented on securesystemslib/keys.py:
```diff
@@ -474,7 +474,7 @@ def format_keyval_to_metadata(keytype, scheme, key_value, private=False):
 
 
-def format_metadata_to_key(key_metadata):
+def format_metadata_to_key(key_metadata, hash_algorithms=securesystemslib.settings.HASH_ALGORITHMS):
```
Addressed in a follow on commit.

joshuagl commented on securesystemslib/keys.py:
```diff
@@ -518,6 +519,10 @@ def format_metadata_to_key(key_metadata, hash_algorithms=securesystemslib.settin
        'keyval': {'public': '...',
                   'private': '...'}}
 
+    hash_algorithms:
+      An optional parameter used to provide custom hash algorithms when
+      creating the key dictionary.
+
```
One purpose of the docstring is to prevent the reader from having to read the code to understand how the function works and what arguments to pass. Therefore it should tell us the type of the argument, especially where it isn't obvious from the function signature.

```suggestion
    hash_algorithms:
      An optional list containing the names of hash algorithms to use when
      generating keyids. Defaults to securesystemslib.settings.HASH_ALGORITHMS

```

MVrachev replied:
Addressed in a follow on commit.

MVrachev commented on tests/test_keys.py:
```diff
@@ -186,7 +186,7 @@ def test_format_metadata_to_key(self):
     self.assertRaises(TypeError, KEYS.format_metadata_to_key)
     args = (self.rsakey_dict, self.rsakey_dict)
-    self.assertRaises(TypeError, KEYS.format_metadata_to_key, *args)
+    self.assertRaises(TypeError, ['sha256', 'sha512'], KEYS.format_metadata_to_key, *args)
```
> Looks like asserting for a `TypeError` is not enough here. Your are getting a `TypeError: 'list' object is not callable` , which is not what you want. I recommend to check out the documentation of unittest's `assertRaises` and maybe read up on *args and **kwargs. :)

Yes, that was a dummy mistake.
Thank you!

joshuagl commented on securesystemslib/keys.py:
```diff
@@ -522,6 +522,10 @@ def format_metadata_to_key(key_metadata, default_keyid=None):
       hash algorithm. If provided, the default keyid can be any string.
 
+    hash_algorithms:
+      An optional parameter used to provide custom hash algorithms when
```
I may be over-reading this, but I'm finding the term "custom" to be problematic. It's not a custom hash algorithm, it's choosing which of the supported hash algorithms you want to use.
```suggestion
      An optional list of hash algorithms to use when
```

MVrachev replied:
Yes, not a bad idea and this will provide information about the type of that argument which is good.

I add and squashed my previous commits.

joshuagl commented on securesystemslib/keys.py:
```diff
@@ -474,7 +474,7 @@ def format_keyval_to_metadata(keytype, scheme, key_value, private=False):
 
 
-def format_metadata_to_key(key_metadata, default_keyid=None):
+def format_metadata_to_key(key_metadata, default_keyid=None, hash_algorithms=securesystemslib.settings.HASH_ALGORITHMS):
```
The [coding style](https://github.com/secure-systems-lab/code-style-guidelines) requests we keep lines to 79 characters or less with a double (4 space) indentation for continuation. Could you wrap this line?

```suggestion
def format_metadata_to_key(key_metadata, default_keyid=None,
    hash_algorithms=securesystemslib.settings.HASH_ALGORITHMS):
```

MVrachev replied:
Yes, you are right. It seems I overlooked it.

Fixed and rebased on the master branch.

joshuagl commented:
This change looks good, thank you.

I would like to hold off on merging it until we have a solid release out to support tuf changes related to the Warehouse work for pep 458, because I anticipate that this will be the first of several API changes as we work to remove the settings module (re: #219) and I would like us to include those in one release rather than spread them over multiple releases. 

We just had the 0.15.0 release with abstract files & directories support. Next we'll want a release with the HSM support in #229. Then we'll have all of the features required of PEP 458 and can start landing changes like this towards #219.

It would be really nice if we tested `format_metadata_to_key` with non-default parameters too, do you think you could work on a patch for that in this PR @MVrachev ?

joshuagl commented:
> Also, be aware that default values in functions are only evaluated once when the function definition is evaluated. This has bitten me before (see #202). As a consequence, constructs like these won't be possible anymore.
> 
> ```python
> # The default arg is evaluated together with the function definition on first import:
> # format_metadata_to_key(key_metadata, hash_algorithms=['sha256', 'sha512'])
> import securesystemslib.keys
> import securesystemslib.settings
> 
> securesystemslib.settings.HASH_ALGORITHMS = ["blake2"]
> 
> format_metadata_to_key(key_metadata)
> # the default for hash_algorithms still is ['sha256', 'sha512'] and not ["blake2"]
> ```

I've been wondering what the right pattern for this is. Maybe something like:
```python
def format_metadata_to_key(key_metadata, default_keyid=None,
  keyid_hash_algorithms=None):
  # docstrings and stuff ...
  if keyid_hash_algorithms is None:
    keyid_hash_algorithms = securesystemslib.settings.HASH_ALGORITHMS
  # ... rest of the function
```

That way we don't bind the default value at function evaluation and we allow users to continue with the current pattern of assigning `securesystemslib.settings.HASH_ALGORITHMS`, making this change less of an API break. We're also a little bit cleaner in our intent to deprecate the settings module as we don't encode it in the interface (by using a variable from the module as a default value).

trishankatdatadog commented:
> I've been wondering what the right pattern for this is. Maybe something like:

Good idea, I like it!

MVrachev commented:
I rebased and addressed Joshua's comments.I changed the following: 

1) Renamed has_algorithms to keyid_hash_algorithms 

2) Use the approach proposed from Joshua Lock about handling the default value of keyid_hash_algorithms 

3) Added a new commit adding unit tests using nondefault hash algorithms.

joshuagl commented on tests/test_keys.py:
```diff
@@ -197,6 +197,29 @@ def test_format_metadata_to_key(self):
 
 
 
+  def test_format_metadata_to_key_non_default_hash_algorithms(self):
```
I think we should just roll these tests into `test_format_metadata_to_key()` – that way we keep all tests of this function together, as is the pattern throughout the rest of the tests.

MVrachev replied:
Addressed your comment by making changes and amending them to my last commit.

joshuagl commented on tests/test_keys.py:
```diff
@@ -197,6 +197,29 @@ def test_format_metadata_to_key(self):
+  def test_format_metadata_to_key_non_default_hash_algorithms(self):
+    # Reconfiguring rsakey_dict to conform to KEY_SCHEMA
+    # i.e. {keytype: 'rsa', keyval: {public: pub_key, private: priv_key}}
+    keyid = self.rsakey_dict['keyid']
```
As a matter of taste I would much prefer that, rather than saving the value of the keyid here and restoring it later, we just make a copy of `rsakey_dict` to use within the test. Then we don't have to worry about restoring the class member to its pristine state at the end of the test.

joshuagl commented on tests/test_keys.py:
```diff
@@ -197,6 +197,29 @@ def test_format_metadata_to_key(self):
+    # Check if the format of the object returned by this function corresponds
+    # to RSAKEY_SCHEMA format.
+    self.assertEqual(None,
+        securesystemslib.formats.RSAKEY_SCHEMA.check_match(rsakey_dict_from_meta),
```
Should we use `assertTrue()` and [`matches()`](https://github.com/secure-systems-lab/securesystemslib/blob/09cf6d4fe8c5131944a8cf0f109fae7d14ba29b2/securesystemslib/schema.py#L70), rather than `assertEqual()` and [`check_match()`](https://github.com/secure-systems-lab/securesystemslib/blob/09cf6d4fe8c5131944a8cf0f109fae7d14ba29b2/securesystemslib/schema.py#L86), here? 

`matches()` returns a boolean to indicate whether the schema is matched, whereas `check_match()` returns `None` on a match and throws an exception on a mismatch. I don't think we want tests throwing unhandled exceptions and testing for a `None`-value or an `Exception` requires more code.

MVrachev replied:
That's a good suggestion. I will add it.

joshuagl commented on tests/test_keys.py:
```diff
@@ -197,6 +197,29 @@ def test_format_metadata_to_key(self):
+        FORMAT_ERROR_MSG)
+
+    self.assertEqual(None,
+        securesystemslib.formats.KEY_SCHEMA.check_match(rsakey_dict_from_meta),
```
Same question on `matches()` vs `check_match()` here.

joshuagl commented on tests/test_keys.py:
```diff
@@ -197,6 +197,29 @@ def test_format_metadata_to_key(self):
+        securesystemslib.formats.KEY_SCHEMA.check_match(rsakey_dict_from_meta),
+        FORMAT_ERROR_MSG)
+
+    self.rsakey_dict['keyid'] = keyid
```
As above, we can avoid this if we make a local copy of the dict to use during this test.

MVrachev replied:
Addressed your comment by making changes and amending them to my last commit.

joshuagl commented on tests/test_keys.py:
```diff
@@ -164,36 +165,50 @@ def test_import_rsakey_from_public_pem(self):
-    del self.rsakey_dict['keyid']
+    # Copying self.rsakey_dict so that rsakey_dict remains
+    # unchanged during and after this test execution.
+    test_rsakey_dict = copy.deepcopy(self.rsakey_dict)
```
Nit: I think `copy` would suffice here?

MVrachev replied:
Yes, you are right. 

I overdo it a little to make sure the right values were copied but forgot to try with `copy.copy`.

I made the change and amended.

lukpueh commented on securesystemslib/keys.py:
```diff
@@ -474,7 +474,8 @@ def format_keyval_to_metadata(keytype, scheme, key_value, private=False):
 
-def format_metadata_to_key(key_metadata, default_keyid=None):
+def format_metadata_to_key(key_metadata, default_keyid=None,
+  keyid_hash_algorithms=None):
```
Minor nit: Please use double-indentation (4 spaces) for line continuation.

lukpueh commented on tests/test_keys.py:
```diff
@@ -164,36 +165,50 @@ def test_import_rsakey_from_public_pem(self):
+
+    # Call format_metadata_to_key by using custom value for keyid_hash_algorithms
+    rsakey_dict_from_meta_custom, junk = KEYS.format_metadata_to_key(test_rsakey_dict,
+      keyid_hash_algorithms=['sha384'])
```
See comment above about line continuation.

MVrachev commented:
Fixed the indentation.

MVrachev commented:
Thanks for merging that @joshuagl!
I will update the corresponding pr https://github.com/theupdateframework/tuf/pull/1016 in tuf next week.
